### ISSUE TITLE
Why do some function take a buffer-id instead of buffer? | Allow non-literal data in nyxt: URLs

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#2268]
@aartaka I see that you've changed show-url-qrcode to take a buffer-id instead of a buffer.

Why is it so?  It makes it rather inconvenient / inconsistent to call it.

**Comment by aartaka** [comment]
This is because `nyxt:` URLs only allow passing literal/constant/scalar data in them. We cannot pass whole buffers (well, we can, but there's no easy way for that yet) as URL query arguments, can we?

While this is a restriction of `nyxt:` URLs, and a grave one, I have an idea for how we can circumvent that. We need to pass object by-reference in internal page commands. To do that, we just need to check the data passed to the function for being non-constant, store it in `nyxt::*inspected-values*` (should we rather put it in debug.lisp? browser.lisp? it's too useful to stay inspector-only...) and, when accessing the values in `nyxt:` URLs somehow infer that the passed value is a reference/ID of an object in `nyxt::*inspected-values*`.

**Comment by Ambrevar** [comment]
Oh yeah, that'd be much, much better! 

The current implementation forces inconsistent code, like buffer ID but also string URLs.

Can you send a patch?

**Comment by aartaka** [comment]
Will do, once I have time :)

**Comment by Ambrevar** [comment]
My proposal to fix this (I can send a patch if you don't have time, no worries) is to use the same strategy as cl-webkit calbacks: instead of encoding the form into the URL, just encode a reference to a unique (gensymed) callback stored in a global table.

In other words (untested):

```lisp
(defpsmacro lisp-eval (form &optional callback)
  "Request the lisp: URL and invoke callback when there's a successful result."
  (let ((request-id (gensym)))
    `(let ((request-lambda (lambda () ,@form)))
       (setf (gethash ,request-id *lisp-url-callbacks*) ,request-lambda)
       (let ((request (fetch (lisp (format nil "lisp://~a" ,request-id)) ; TODO: Maybe add purely informative string in the title, to make HTML inspection more useful?
                             (create :mode "no-cors"))))
         (when ,callback
           (chain request
                  (then (lambda (response)
                          (when (@ response ok)
                            (chain response (json)))))
                  (then ,callback)))))))


;...

(define-internal-scheme "lisp"
    (lambda (url buffer)
      (let ((url (quri:uri url)))
        ;; TODO: Replace this condition with `(network-buffer-p buffer)`?
        (if (or (status-buffer-p buffer)
                (panel-buffer-p buffer)
                (prompt-buffer-p buffer)
                (internal-url-p (url buffer)))
            (let* ((request-id (make-symbol "G" (quri:uri-host url)))) ; TODO: Replace "G" with something meaningful.
              (log:debug "Evaluate Lisp code from internal page: ~a" request-id)
              (unwind-protect
                   (values (let ((result (funcall (gethash request-id *lisp-url-callbacks*))))
                             ;; Objects and other complex structures make cl-json choke.
                             ;; TODO: Maybe encode it to the format that `cl-json'
                             ;; supports, then we can override the encoding and
                             ;; decoding methods and allow arbitrary objects (like
                             ;; buffers) in the nyxt:// URL arguments..
                             (cl-json:encode-json-to-string
                              (when (or (scalar-p result)
                                        (and (sequence-p result)
                                             (every #'scalar-p result)))
                                result)))
                           "application/json")
                (remhash request-id *lisp-url-callbacks*)))
            (values "undefined" "application/json;charset=utf8"))))
  :cors-enabled-p t
  :error-callback (lambda (c) (log:debug "Error when evaluating lisp URL: ~a" c)))
```

**Comment by Ambrevar** [comment]
Benefits: no more need for quoted code, plus we have closures now!
Drawbacks: none?

@aartaka If you think this is sane, I'll send a patch tomorrow.

**Comment by Ambrevar** [comment]
Err... this only fixes lisp-urls actually!

For Nyxt URLs, the problem of passing by reference is that these Nyxt URLs can be hosted on third-party website, and thus would allow an attack to dereference undesired values.

Another problem is that Nyxt URLs are supposed to be perennial.  Say I open buffer-history-tree on buffer 1234, the URL is something like `nyxt://buffer-history-tree%XYZ%1234`, then I can copy the URL.  If I paste it in `set-url`, it works.
But for references to work here too, it means that we should never collect them, thus preventing garbage collection.

Sooo...  Maybe the solution then is to decouple the Nyxt URL function arguments from the internal page command argument.  It's probably hard to do this automatically (e.g. infer `(id buffer)` from a `buffer` object), so instead we could separate the two:

```lisp

(define-internal-page buffer-history-tree (&key (id (id (current-buffer))))
    (output-buffer (format nil "*History-~a*" id)
                   'nyxt/history-tree-mode:history-tree-mode)
  "Display the history tree of a buffer."
  (let ((buffer (nyxt::buffers-get id))
        (mode (find-submode 'nyxt/history-tree-mode:history-tree-mode output-buffer)))
    (spinneret:with-html-string
      (:style (style mode))
      (:h1 (format nil "History of ~a" buffer))
      (:div (:raw (with-current-buffer buffer (render-buffer-history-tree buffer)))))))

(define-command-global buffer-history-tree (&key (buffer (current-buffer)))
  "Display the history tree of a buffer."
  ;; `internal-page-set' creates the buffer if necessary, then call `nyxt-url'
  ;; over the arguments and `buffer-load' the result.
  (internal-page-set 'buffer-history-tree (id buffer)))
```

This would also fix the issue of being able to create anonymous internal pages.

**Comment by Ambrevar** [comment]
Fixed in #2306.

**Comment by Ambrevar** [comment]
Done with #2354.
